src/components/OnThisPageNav.vue

```diff
@@ -0,0 +1,160 @@
+<!--
+  This source file is part of the Swift.org open source project
+
+  Copyright (c) 2021-2022 Apple Inc. and the Swift project authors
+  Licensed under Apache License v2.0 with Runtime Library Exception
+
+  See https://swift.org/LICENSE.txt for license information
+  See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+-->
+<template>
+  <div class="OnThisPageNav">
+    <slot
+      :active="currentPageSection"
+      :sections="onThisPageSections"
+      :checkIsActive="checkIsActive"
+      :scrollTo="scrollTo"
+      :getItemClasses="getItemClasses"
+    >
+      <ul class="items">
+        <li
+          v-for="item in onThisPageSections"
+          :key="item.anchor"
+          :class="getItemClasses(item)"
+        >
+          <router-link
+            :to="item.url"
+            class="base-link"
+            @click.native="scrollTo(item)"
+          >
+            {{ item.title }}
+          </router-link>
+        </li>
+      </ul>
+    </slot>
+  </div>
+</template>
+
+<script>
+import throttle from 'docc-render/utils/throttle';
+import { waitFrames } from 'docc-render/utils/loading';
+import ScrollToElement from 'docc-render/mixins/scrollToElement';
+import { buildUrl } from 'docc-render/utils/url-helper';
+
+export default {
+  name: 'OnThisPageNav',
+  mixins: [ScrollToElement],
+  inject: {
+    store: {
+      default() {
+        return {
+          state: { onThisPageSections: [], currentPageSection: {} },
+        };
+      },
+    },
+  },
+  computed: {
+    onThisPageSections: ({ store, $route }) => store.state.onThisPageSections.map(item => ({
+      ...item,
+      url: buildUrl(`#${item.anchor}`, $route.query),
+    })),
+    currentPageSection: ({ store }) => store.state.currentPageSection || {},
+  },
+  async mounted() {
+    window.addEventListener('scroll', this.onScroll, false);
+    this.$once('hook:beforeDestroy', () => {
+      window.removeEventListener('scroll', this.onScroll);
+    });
+    // make sure everything is ready
+    await waitFrames(8);
+    this.onScroll();
+  },
+  methods: {
+    onScroll: throttle(function onScroll() {
```

Comment:
It would have, but the IntersectionObserver was not getting fired as fast as I needed it. If you scroll fast, then it just gets throttled and the highlighting is often off.

Reply:
Interesting. I'm pretty sure that the tutorials overview page has a very similar system for highlighting the current section in a table of contents using the observer, but maybe there are differences here that make it not work as well.

Reply:
Yes, that was my initial thought too, how come it works fine there??? 🤔 

But tutorials are a bit different because they are very long sections that hold IDs, but here we have a mix of sections and headings that have the id on them (soon only headings https://github.com/apple/swift-docc-render/pull/408), and I found it would just entirely skip over the heading as it flies by 🤣